crates/valence_nbt/src/value.rs

```diff
@@ -134,63 +135,82 @@ macro_rules! impl_value {
             fn from(v: $($reference)* i64) -> Self {
                 Self::Long(v)
             }
         }
 
-        impl $(<$lifetime>)? From<$($reference)* f32> for $name $(<$lifetime>)? {
+        impl <$($lifetime,)? S> From<$($reference)* f32> for $name<$($lifetime,)? S> {
             fn from(v: $($reference)* f32) -> Self {
                 Self::Float(v)
             }
         }
 
-        impl $(<$lifetime>)? From<$($reference)* f64> for $name $(<$lifetime>)? {
+        impl <$($lifetime,)? S> From<$($reference)* f64> for $name<$($lifetime,)? S> {
             fn from(v: $($reference)* f64) -> Self {
                 Self::Double(v)
             }
         }
 
-        impl $(<$lifetime>)? From<$($reference)* List> for $name $(<$lifetime>)? {
-            fn from(v: $($reference)* List) -> Self {
+        impl <$($lifetime,)? S> From<$($reference)* List<S>> for $name<$($lifetime,)? S> {
+            fn from(v: $($reference)* List<S>) -> Self {
                 Self::List(v)
             }
         }
 
-        impl $(<$lifetime>)? From<$($reference)* Compound> for $name $(<$lifetime>)? {
-            fn from(v: $($reference)* Compound) -> Self {
+        impl <$($lifetime,)? S> From<$($reference)* Compound<S>> for $name<$($lifetime,)? S> {
+            fn from(v: $($reference)* Compound<S>) -> Self {
                 Self::Compound(v)
             }
         }
+
+        impl <$($lifetime,)? S> PartialEq<Self> for $name<$($lifetime,)? S> where S: Eq + Ord + Hash {
+            fn eq(&self, other: &Self) -> bool {
+                match self {
+                    Self::Byte(v) => matches!(other, Self::Byte(other_v) if v == other_v),
+                    Self::Short(v) => matches!(other, Self::Short(other_v) if v == other_v),
+                    Self::Int(v) => matches!(other, Self::Int(other_v) if v == other_v),
+                    Self::Long(v) => matches!(other, Self::Long(other_v) if v == other_v),
+                    Self::Float(v) => matches!(other, Self::Float(other_v) if v == other_v),
+                    Self::Double(v) => matches!(other, Self::Double(other_v) if v == other_v),
+                    Self::ByteArray(v) => matches!(other, Self::ByteArray(other_v) if v == other_v),
+                    Self::String(v) => matches!(other, Self::String(other_v) if v == other_v),
+                    Self::List(v) => matches!(other, Self::List(other_v) if v == other_v),
+                    Self::Compound(v) => matches!(other, Self::Compound(other_v) if v == other_v),
+                    Self::IntArray(v) => matches!(other, Self::IntArray(other_v) if v == other_v),
+                    Self::LongArray(v) => matches!(other, Self::LongArray(other_v) if v == other_v),
+                }
+            }
+        }
     }
 }
 
 impl_value!(Value,,(*),);
 impl_value!(ValueRef, 'a, (**), &'a);
 impl_value!(ValueMut, 'a, (**), &'a mut);
 
-impl Value {
+impl<S> Value<S> {
     /// Converts a reference to a value to a [ValueRef].
```

Comment:
```suggestion
    /// Converts a reference to a value to a [`ValueRef`].
```

Reply:
Looks like there's a lot of instances of this. Mind if I fix them all in bulk in a separate PR?

Reply:
Yeah that's fine